src/topology/algebra/nonarchimedean/basic.lean

```diff
@@ -0,0 +1,138 @@
+nonarchimedean_add_group R := {..t}
+
+namespace nonarchimedean_group
+
+variables {G : Type*} [group G] [topological_space G] [nonarchimedean_group G]
+variables {H : Type*} [group H] [topological_space H] [topological_group H]
+variables {K : Type*} [group K] [topological_space K] [nonarchimedean_group K]
+
+/-- If a topological group embeds into a nonarchimedean group, then it
+  is nonarchimedean. -/
+@[to_additive nonarchimedean_add_group.nonarchimedean_of_emb]
+lemma nonarchimedean_of_emb (f : G →* H) (emb : open_embedding f) : nonarchimedean_group H :=
+{ is_nonarchimedean := λ U hU, have h₁ : (f ⁻¹' U) ∈ nhds (1 : G), from
+    by {apply emb.continuous.tendsto, rwa is_group_hom.map_one f},
+  let ⟨V, hV⟩ := is_nonarchimedean (f ⁻¹' U) h₁ in
+    ⟨{is_open' := emb.is_open_map _ V.is_open, ..subgroup.map f V},
+      set.image_subset_iff.2 hV⟩ }
+
+/-- An open neighborhood of the identity in the cartesian product of two nonarchimedean groups
+  contains the cartesian product of an open neighborhood in each group. -/
+@[to_additive nonarchimedean_add_group.prod_subset]
+lemma prod_subset {U} (hU : U ∈ nhds (1 : G × K)) :
+  ∃ (V : open_subgroup G) (W : open_subgroup K), (V : set G).prod (W : set K) ⊆ U :=
+begin
+  erw [nhds_prod_eq, filter.mem_prod_iff] at hU,
+  rcases hU with ⟨U₁, hU₁, U₂, hU₂, h⟩,
+  cases is_nonarchimedean _ hU₁ with V hV,
+  cases is_nonarchimedean _ hU₂ with W hW,
+  use V, use W,
+  rw set.prod_subset_iff,
+  intros x hX y hY,
+  refine set.subset.trans (set.prod_mono hV hW) h _,
+  exact set.mem_sep hX hY,
+end
+
+/-- An open neighborhood of the identity in the cartesian square of two nonarchimedean groups
+  contains the cartesian product of an open neighborhood in each group. -/
+@[to_additive nonarchimedean_add_group.prod_self_subset]
+lemma prod_self_subset {U} (hU : U ∈ nhds (1 : G × G)) :
+  ∃ (V : open_subgroup G), (V : set G).prod (V : set G) ⊆ U :=
+let ⟨V, W, h⟩ := prod_subset hU in
+  ⟨V ⊓ W, by {refine set.subset.trans (set.prod_mono _ _) ‹_›; simp}⟩
+
+/-- The cartesian product of two nonarchimedean groups is nonarchimedean. -/
+@[to_additive]
+instance : nonarchimedean_group (G × K) :=
+{ is_nonarchimedean := λ U hU, let ⟨V, W, h⟩ := prod_subset hU in ⟨V.prod W, ‹_›⟩ }
+
+end nonarchimedean_group
+
+namespace nonarchimedean_ring
+
+open nonarchimedean_ring
+open nonarchimedean_add_group
+
+variables {R S : Type*}
+variables [ring R] [topological_space R] [nonarchimedean_ring R]
+variables [ring S] [topological_space S] [nonarchimedean_ring S]
+
+/-- The cartesian product of two nonarchimedean rings is nonarchimedean. -/
+instance : nonarchimedean_ring (R × S) :=
+{ is_nonarchimedean := nonarchimedean_add_group.is_nonarchimedean }
+
+/-- If you multiply an element of a nonarchimedean ring by an open subgroup,
+  the product still contains an open subgroup. -/
+lemma left_mul_subset (U : open_add_subgroup R) (r : R) :
+  ∃ V : open_add_subgroup R, r • (V : set R) ⊆ U :=
```

Comment:
The doc-string seems to say something different than the lemma.

Reply:
Fixed